### PR TITLE
[metric] Add a new metrics DISTINCT.

### PR DESCRIPTION
**Pull request by xiemoyuan**
### PR types
New features

### PR changes
APIs

### Describe
Add a new metric DISTINCT to evaluate the divesity of generate text.

### CODE REVIEW

File: paddlenlp/metrics/distinct.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            import paddle
+            import paddle.nn.functional as F
+            from paddlenlp.transformers import BertTokenizer
+            from paddlenlp.metrics import Distinct
+            from functools import partial
+            def trans_func(logits, tokenizer):
+                '''Transform the network output `logits` to string list.'''
+                probs = F.softmax(logits)
+                # [batch_size, seq_len]
+                token_ids = paddle.argmax(probs, axis=-1).numpy().tolist()
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: 如果是只是计算`argmax`的话，前面是否不需要计算`softmax`呢

File: paddlenlp/metrics/distinct.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+__all__ = ['Distinct']
+
+
+def _is_numpy_(var):
+    return isinstance(var, (np.ndarray, np.generic))
+
+
+class Distinct(paddle.metric.Metric):
+    r"""
+    Distinct is an algorithm for evaluating the quality of text which has been
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: 能帮忙在[metrics.md](https://github.com/PaddlePaddle/PaddleNLP/blob/release/2.0-rc/docs/metrics.md)中为该新增API添加一句简单的介绍吗，感谢

File: paddlenlp/metrics/distinct.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        2. Using as an instance of `paddle.metric.Metric`.
+                
+        .. code-block:: python
+
+            import paddle
+            import paddle.nn.functional as F
+            from paddlenlp.transformers import BertTokenizer
+            from paddlenlp.metrics import Distinct
+            from functools import partial
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: `from functools import partial`是否应该在最前面import

File: paddlenlp/metrics/distinct.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    return isinstance(var, (np.ndarray, np.generic))
+
+
+class Distinct(paddle.metric.Metric):
+    r"""
+    Distinct is an algorithm for evaluating the quality of text which has been 
+    generated. See detail at https://arxiv.org/abs/1510.03055
+
+    Distinct could be used as `paddle.metric.Metric` class, or an ordinary 
+    class. When Distinct is used as `paddle.metric.Metric` class. A function is
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: Distinct这个类名在doc中是否需要用\`符号包围呢 @guoshengCS

File: paddlenlp/metrics/distinct.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        Parameters:
+            probs: The generate token probability, with shape 
+                [batch_size, seq_len, vocab_size].
+        """
+        if _is_numpy_(logits):
+            logits = paddle.to_tensor(logits)
+
+        assert self.trans_func is not None, "The `update` method requires user "\
+            "to provide `trans_func` when initializing Distinct."
+        cand_list = self.trans_func(logits)
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: 前面这部分关于tensor计算逻辑就放在`compute`里吧。`update`中一般是做`numpy.ndarray`的计算。

File: paddlenlp/metrics/distinct.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        Update the metrics states.
+        
+        Parameters:
+            probs: The generate token probability, with shape 
+                [batch_size, seq_len, vocab_size].
+        """
+        if _is_numpy_(logits):
+            logits = paddle.to_tensor(logits)
+
+        assert self.trans_func is not None, "The `update` method requires user "\
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py:100]: 既然这里不允许`self.trans_func`为None，就在`__init__`函数里把默认参数去掉吧。另外你那里自定义的`trans_func`如果有通用性的话，能否也一起提供呢

File: paddlenlp/metrics/distinct.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+__all__ = ['Distinct']
+
+
+def _is_numpy_(var):
+    return isinstance(var, (np.ndarray, np.generic))
+
+
+class Distinct(paddle.metric.Metric):
+    r"""
```
Comment (LiuChiachi) [inline paddlenlp/metrics/distinct.py]: 文档中如果没有数学公式貌似不需要r，不过这个可能也会看以后规范

File: paddlenlp/metrics/distinct.py (hunk 6 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py:100]: 当这个metric作为普通用法时（输入字符串tokens），是不需要trans_func的

File: paddlenlp/metrics/distinct.py (hunk 1 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: Done. Thanks.

File: paddlenlp/metrics/distinct.py (hunk 2 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: Done. Thanks.

File: paddlenlp/metrics/distinct.py (hunk 3 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: Done. Thanks.

File: paddlenlp/metrics/distinct.py (hunk 4 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: Done. Thanks.

File: paddlenlp/metrics/distinct.py (hunk 5 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: 已修改

File: paddlenlp/metrics/distinct.py (hunk 6 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py:100]: example里定义的`trans_func`不具有通用性

File: paddlenlp/metrics/distinct.py (hunk 7 above)
Comment (xiemoyuan) [reply paddlenlp/metrics/distinct.py]: Done. Thanks.

File: paddlenlp/metrics/distinct.py (hunk 6 above)
Comment (LiuChiachi) [reply paddlenlp/metrics/distinct.py:100]: 好的

**Review by LiuChiachi — DISMISSED** [review]
LGTM

File: paddlenlp/metrics/distinct.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+import numpy as np
+import paddle
+
+__all__ = ['Distinct']
+
+
+class Distinct(paddle.metric.Metric):
+    """
+    Distinct is an algorithm for evaluating the quality of text which has been 
+    generated. See detail at https://arxiv.org/abs/1510.03055
```
Comment (guoshengCS) [inline paddlenlp/metrics/distinct.py]: > an algorithm for evaluating the quality of text which has been

感觉这里介绍的有点略泛了，感觉上可以是任何文本生成评估指标的描述，最好稍微描述下Distinct。

File: paddlenlp/metrics/distinct.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    def __init__(self, n_size=2, trans_func=None, name="distinct"):
+        super(Distinct, self).__init__()
+        self._name = name
+        self.diff_ngram = set()
+        self.count = 0.0
+        self.n_size = n_size
+        self.trans_func = trans_func
+
+    def update(self, logits, *args):
```
Comment (guoshengCS) [inline paddlenlp/metrics/distinct.py]: 感觉上可能最好先把logits参数换成output，想着如果后续`token_ids = np.argmax(logits, axis=-1)`慢考虑性能问题的话可能增加compute方法使用paddle.tensor来进行argmax，这个输入就不一定叫做logits了，`self.trans_func`只要处理token_ids就行（用户也可以像当前这样不使用compute），这个在文档里说明`update会对output使用trans_func进行处理返回candidate list，逐个candidate经过add_inst `就好

File: paddlenlp/metrics/distinct.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Update the metrics states.
+        
+        Args:
+            probs: The generate token probability, with shape
```
Comment (guoshengCS) [inline paddlenlp/metrics/distinct.py]: 注意文档和代码的一致性，另外如果args暂时没有用的话要不就先去掉呢
